Popover: lock the __experimentalPopoverPositionToPlacement function

ciampo opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->

Part of #47196
**Requires #47229 to be merged first**

In an effort to reduce the amount of experimental APIs introduced to the repo, this PR uses [the experimental `lock`/`unlock` APIs](https://github.com/WordPress/gutenberg/pull/46131) to lock the `__experimentalPopoverPositionToPlacement` function from the `@wordpress/components` package, renaming it to `__experimentalPopoverLegacyPositionToPlacement `.

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

I had initially decided to stabilise this APIs, but following the conversations in this PR I later changed the approach and went for `lock` / `unlock`. The whole point here is to prevent the `__experimentalPopoverPositionToPlacement` from being released to core and therefore losing its "experimental" nature.

Also, I decided not to introduce any soft deprecations and directly rename the export because it doesn't look like [there any usages according to WPDirectory](https://wpdirectory.net/search/01GQSVMVKQDRA8QV605BVQM858).

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

- Removed the original export from `@wordpress/components`, and instead added the function under the `experiments` object (which is locked)
- Refactored the consumers (`URLPopover`) to use the `unlock` APIs to retrieve the same function

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

- Make sure project build
- Make sure `Popover` Storybook example work as expected
- Make sure that the `URLPopover` is positioned correctly in the editor (as on `trunk`)
- Make sure that there aren't any leftover mentions of the `__experimentalPopoverPositionToPlacement` prop

<details>

<summary> Detailed instructions on how to test the `URLPopover` component</summary>

1. Apply this diff in order to use the legacy `position` prop, which should trigger the unlocked function

```diff
diff --git a/packages/block-editor/src/components/url-popover/image-url-input-ui.js b/packages/block-editor/src/components/url-popover/image-url-input-ui.js
index 57084ae76c..aa51c4ba58 100644
--- a/packages/block-editor/src/components/url-popover/image-url-input-ui.js
+++ b/packages/block-editor/src/components/url-popover/image-url-input-ui.js
@@ -252,6 +252,7 @@ const ImageURLInputUI = ( {
 			/>
 			{ isOpen && (
 				<URLPopover
+					position="top"
 					anchor={ popoverAnchor }
 					onFocusOutside={ onFocusOutside() }
 					onClose={ closeLinkUI }
```

2. In the post editor, insert an image block and add an image to it
3. Select the image, and in the block toolbar click on the "Insert link" button
4. Make sure that the resulting `URLPopover` renders on top of the block toolbar, and that a warning is printed to the browser's console (since `position` is deprecated)

ciampo committed:
Rename and stabilize __experimentalPopoverPositionToPlacement

ciampo committed:
Rename internal tests import statement

ciampo commented:
Marking as in need of dev note in case we establish that a prop rename is equivalent to a new prop

talldan commented:
I'm thinking that it might be better to avoid stabilizing this function. The cons are that It's a legacy function but will become a stable API indefinitely once released in WordPress core. It will have very limited use as everyone switches over to `placement`.

There seems to only be one use of the function outside of the components package, so other options could be either removing the export and duplicating the function or using the lock/unlock experiments API.

aaronrobertshaw reviewed (commented):
> The cons are that It's a legacy function but will become a stable API indefinitely once released in WordPress core

This is a good point. 

I don't have a strong preference towards either of the options @talldan suggested. Either would make it easier to remove the function in the longer term.

Other than that:

✅ Smoke test in editors didn't reveal any issues
✅ Unit tests pass & no typing errors

tyxla reviewed (commented):
I wonder if, in line with what @talldan and @aaronrobertshaw said, it makes sense to deprecate all APIs that are related to the old popover position prop, instead of stabilizing this API? In an ideal world where we're getting rid of deprecated APIs after some period, we'll be able to clean the legacy code up, instead of having it with us forever. Is there a good reason to want to keep maintaining the legacy API?

ciampo committed:
Revert "Rename internal tests import statement"

This reverts commit 4a0602cc3cb73c045fd57dec2952727eed4f446b.

ciampo committed:
Revert "Rename and stabilize __experimentalPopoverPositionToPlacement"

This reverts commit 0d0a7aa7d0a27c40f6f13331155b08e2c0ff2840.

ciampo committed:
Use lock/unlock APIs to lock the `positionToPlacement` function

ciampo committed:
Update CHANGELOG

ciampo commented:
Thanks everyone for the feedback!

I think that using `lock` / `unlock` may be the right call here — we'll be able to clean it up when the `position` prop will be fully deprecated and removed from the codebase.

Therefore, I went ahead and:
- reverted the early commits in this PR
- rebased on top of #47229 to include the great work that @noisysocks did there
- locked/unlocked the `posisionToPlacement` prop
- updated the PR's description and added more details testing instructions

Feel free to give this PR a new round of reviews. I'll merge this PR only after #47229 gets merged.

noisysocks commented on packages/components/CHANGELOG.md:
```diff
@@ -24,6 +24,7 @@
 -   `Button`: Convert to TypeScript ([#46997](https://github.com/WordPress/gutenberg/pull/46997)).
 -   `QueryControls`: Convert to TypeScript ([#46721](https://github.com/WordPress/gutenberg/pull/46721)).
 -   `Notice`: refactor to TypeScript ([47118](https://github.com/WordPress/gutenberg/pull/47118)).
+-   Lock the `__experimentalPopoverPositionToPlacement` function and rename it to `popoverLegacyPositionToPlacement` ([#47505](https://github.com/WordPress/gutenberg/pull/47505)).
```
I think we ended up deciding in https://github.com/WordPress/gutenberg/pull/47229 that it's nice to keep the `__experimental` prefix so that can quickly identify which APIs are non-public.

Or maybe we could use a new `__private` prefix for these cases?

I saw @youknowriad discussing these sorts of things in Slack. What do you think?

When the dust settles on 6.2 we should update `backward-compatibility.md` with the best practices that we land on.

noisysocks replied:
Oh, but this is a function! not a prop. OK then I guess that's fine. Maybe. I don't know anymore 😭

ciampo replied:
Added the `__experimental` prefix for good measure (https://github.com/WordPress/gutenberg/pull/47505/commits/9f3158b79c741d6aa29d989a7f8994f89f3dafea). As you say, we can align naming conventions after 6.2 is released

ciampo committed:
Add __experimental prefix

ciampo committed:
Update CHANGELOG

ciampo commented:
With https://github.com/WordPress/gutenberg/pull/47229 merged, I rebased this PR to include the latest changes from `trunk`.

This PR should be now ready for a final round of review

aaronrobertshaw reviewed (approved):
Thanks for iterating on this one @ciampo 👍 

It's testing well for me.

✅ Project builds successfully
✅ Storybook examples work as expected
✅ URLPopover is positioned correctly and deprecation warning is output when using `position` prop
✅ No remaining mentions of `__experimentalPopoverPositionToPlacement` prop

LGTM! 🚢

tyxla reviewed (approved):
Nice work 👍 

Needs a rebase, but the code looks great and it tests well.

🚀

ntsekouras commented:
Cherry-picked this PR to the wp/6.2 branch.
